File: Assets/Scripts/Input/InputParser.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using UnityEngine;

namespace Cristal.CLI.Input
{
    /// <summary>
    /// Parses player input into structured commands and semantic signals.
    /// Supports commands with arguments like "invoke arcana XIII" or "read /core/memory".
    /// </summary>
    public static class InputParser
    {
        // Regex patterns for command parsing
        private static readonly Regex CommandPattern = new Regex(
            @"^(\w+)\s*(.*?)$",
            RegexOptions.Compiled | RegexOptions.IgnoreCase
        );

        private static readonly Regex PathArgumentPattern = new Regex(
            @"(/[\w/\.\-]+)",
            RegexOptions.Compiled
        );

        // Emotional keyword sets
        private static readonly HashSet<string> PositiveKeywords = new HashSet<string>
        {
            "hope", "love", "happy", "joy", "peace", "light", "warm", "trust", "beautiful", "good", "yes", "thank"
        };

        private static readonly HashSet<string> NegativeKeywords = new HashSet<string>
        {
            "fear", "hate", "scared", "afraid", "alone", "lost", "dark", "pain", "cold", "empty", "dead", "sad", "angry"
        };

        private static readonly HashSet<string> PhilosophicalKeywords = new HashSet<string>
        {
            "why", "meaning", "purpose", "truth", "real", "exist", "life", "death", "soul", "consciousness"
        };

        private static readonly HashSet<string> IdentityKeywords = new HashSet<string>
        {
            "who", "what", "am", "identity", "name", "self"
        };

        private static readonly HashSet<string> MemoryKeywords = new HashSet<string>
        {
            "remember", "memory", "recall", "past", "before", "forgot", "history"
        };

        private static readonly HashSet<string> RitualKeywords = new HashSet<string>
        {
            "invoke", "summon", "arcana", "ritual", "activate", "awaken", "call"
        };

        /// <summary>
        /// Parse raw input into a structured ParsedCommand.
        /// </summary>
        public static ParsedCommand Parse(string input)
        {
            var result = new ParsedCommand
            {
                Raw = input ?? "",
                Keywords = new List<string>()
            };

            // Handle empty/whitespace input
            if (string.IsNullOrWhiteSpace(input))
            {
                result.IsSemanticSignal = true;
                result.SignalType = SemanticSignalType.Empty;
                return result;
            }

            string trimmed = input.Trim();
            string lower = trimmed.ToLower();

            // Extract keywords
            result.Keywords = ExtractKeywords(trimmed);

            // Calculate emotional weight
            result.EmotionalWeight = CalculateEmotionalWeight(lower);

            // Check if it's a command structure
            var match = CommandPattern.Match(trimmed);
            if (match.Success)
            {
                string potentialCommand = match.Groups[1].Value.ToLower();
                string argumentPart = match.Groups[2].Value.Trim();

                if (KnownCommands.IsKnownCommand(potentialCommand))
                {
                    result.IsCommand = true;
                    result.Command = potentialCommand;
                    result.ArgumentString = argumentPart;
                    result.Arguments = ParseArguments(argumentPart);
                    result.SignalType = SemanticSignalType.None;
                    return result;
                }
            }

            // Not a command - classify as semantic signal
            result.IsCommand = false;
            result.IsSemanticSignal = true;
            result.SignalType = ClassifySemanticSignal(lower, trimmed);

            return result;
        }

        /// <summary>
        /// Parse argument string into individual arguments.
        /// Handles quoted strings and path-style arguments.
        /// </summary>
        private static string[] ParseArguments(string argumentString)
        {
            if (string.IsNullOrWhiteSpace(argumentString))
            {
                return new string[0];
            }

            var arguments = new List<string>();
            var current = "";
            bool inQuotes = false;

            for (int i = 0; i < argumentString.Length; i++)
            {
                char c = argumentString[i];

                if (c == '"' || c == '\'')
                {
                    inQuotes = !inQuotes;
                }
                else if (c == ' ' && !inQuotes)
                {
                    if (!string.IsNullOrWhiteSpace(current))
                    {
                        arguments.Add(current);
                        current = "";
                    }
                }
                else
                {
                    current += c;
                }
            }

            if (!string.IsNullOrWhiteSpace(current))
            {
                arguments.Add(current);
            }

            return arguments.ToArray();
        }

        /// <summary>
        /// Classify non-command input as a semantic signal type.
        /// </summary>
        private static SemanticSignalType ClassifySemanticSignal(string lower, string original)
        {
            // Check for empty/nonsense
            if (IsNonsense(original))
            {
                return SemanticSignalType.Nonsense;
            }

            // Check for profanity/frustration
            if (ContainsProfanity(lower))
            {
                return SemanticSignalType.Profanity;
            }

            // Check for greetings
            if (IsGreeting(lower))
            {
                return SemanticSignalType.Greeting;
            }

            // Check for farewell/exit
            if (IsFarewell(lower))
            {
                return SemanticSignalType.Farewell;
            }

            // Check for affirmation
            if (IsAffirmation(lower))
            {
                return SemanticSignalType.Affirmation;
            }

            // Check for negation
            if (IsNegation(lower))
            {
                return SemanticSignalType.Negation;
            }

            // Check for identity questions
            if (ContainsAny(lower, IdentityKeywords) && (lower.Contains("?") || lower.Contains("who") || lower.Contains("what")))
            {
                return SemanticSignalType.Identity;
            }

            // Check for memory-related
            if (ContainsAny(lower, MemoryKeywords))
            {
                return SemanticSignalType.Memory;
            }

            // Check for ritual/arcana keywords
            if (ContainsAny(lower, RitualKeywords))
            {
                return SemanticSignalType.Ritual;
            }

            // Check for philosophical
            if (ContainsAny(lower, PhilosophicalKeywords))
            {
                return SemanticSignalType.Philosophical;
            }

            // Check for questions
            if (lower.Contains("?"))
            {
                return SemanticSignalType.Question;
            }

            // Check for emotional content
            if (ContainsAny(lower, PositiveKeywords) || ContainsAny(lower, NegativeKeywords))
            {
                return SemanticSignalType.Emotional;
            }

            return SemanticSignalType.None;
        }

        /// <summary>
        /// Extract meaningful keywords from input.
        /// </summary>
        private static List<string> ExtractKeywords(string input)
        {
            var keywords = new List<string>();
            var stopWords = new HashSet<string>
            {
                "the", "a", "an", "is", "are", "was", "were", "am", "i", "you", "we", "they", "it",
                "to", "of", "and", "or", "in", "on", "at", "for", "with", "do", "does", "did",
                "have", "has", "had", "be", "been", "being", "my", "your", "our", "their"
            };

            string[] words = input.ToLower().Split(
                new char[] { ' ', ',', '.', '!', '?', ';', ':', '"', '\'', '/', '\\' },
                StringSplitOptions.RemoveEmptyEntries
            );

            foreach (string word in words)
            {
                if (word.Length >= 3 && !stopWords.Contains(word))
                {
                    keywords.Add(word);
                }
            }

            return keywords;
        }

        /// <summary>
        /// Calculate emotional weight of input.
        /// </summary>
        private static float CalculateEmotionalWeight(string lower)
        {
            float weight = 0f;

            foreach (string word in PositiveKeywords)
            {
                if (lower.Contains(word)) weight += 0.5f;
            }

            foreach (string word in NegativeKeywords)
            {
                if (lower.Contains(word)) weight -= 0.5f;
            }

            // Intensity multipliers
            if (lower.Contains("!") || lower.Contains("?!")) weight *= 1.2f;
            if (lower.Contains("very") || lower.Contains("so much") || lower.Contains("really")) weight *= 1.3f;
            if (lower.Contains("always") || lower.Contains("never")) weight *= 1.2f;

            return Mathf.Clamp(weight, -2f, 2f);
        }

        private static bool ContainsAny(string text, HashSet<string> keywords)
        {
            foreach (string keyword in keywords)
            {
                if (text.Contains(keyword)) return true;
            }
            return false;
        }

        private static bool IsNonsense(string input)
        {
            // Check if input is mostly non-alphabetic
            int alphaCount = 0;
            foreach (char c in input)
            {
                if (char.IsLetter(c)) alphaCount++;
            }
            return input.Length > 0 && (float)alphaCount / input.Length < 0.3f;
        }

        private static bool ContainsProfanity(string lower)
        {
            // Basic check - could be expanded
            var profanity = new HashSet<string> { "fuck", "shit", "damn", "hell", "ass" };
            foreach (string word in profanity)
            {
                if (lower.Contains(word)) return true;
            }
            return false;
        }

        private static bool IsGreeting(string lower)
        {
            var greetings = new HashSet<string> { "hello", "hi", "hey", "greetings", "hola", "howdy" };
            string firstWord = lower.Split(' ')[0].TrimEnd('!', '.', ',');
            return greetings.Contains(firstWord);
        }

        private static bool IsFarewell(string lower)
        {
            var farewells = new HashSet<string> { "bye", "goodbye", "exit", "quit", "leave", "farewell", "adios" };
            string firstWord = lower.Split(' ')[0].TrimEnd('!', '.', ',');
            return farewells.Contains(firstWord) || lower.Contains("good bye");
        }

        private static bool IsAffirmation(string lower)
        {
            var affirmations = new HashSet<string> { "yes", "ok", "okay", "sure", "continue", "proceed", "go", "si", "yep", "yeah" };
            string trimmed = lower.TrimEnd('!', '.', ',', '?');
            return affirmations.Contains(trimmed) || affirmations.Contains(lower.Split(' ')[0]);
        }

        private static bool IsNegation(string lower)
        {
            var negations = new HashSet<string> { "no", "stop", "cancel", "don't", "dont", "never", "nope" };
            string trimmed = lower.TrimEnd('!', '.', ',', '?');
            return negations.Contains(trimmed) || negations.Contains(lower.Split(' ')[0]);
        }
    }
}

File: web/src/terminal/inputParser.ts
import { ParsedCommand, SemanticSignalType } from "./types";

// Port of Cristal.CLI.Input.InputParser.

const KNOWN_COMMANDS = new Set<string>([
  "help", "status", "clear", "reset", "save", "load", "export",
  "read", "list", "show", "view",
  "see", "visions", "vision",
  "invoke", "summon", "activate", "unlock",
  "remember", "recall", "forget", "log",
  "echo", "corrupt", "stabilize", "seek",
]);

const POSITIVE = new Set(["hope", "love", "happy", "joy", "peace", "light", "warm", "trust", "beautiful", "good", "yes", "thank"]);
const NEGATIVE = new Set(["fear", "hate", "scared", "afraid", "alone", "lost", "dark", "pain", "cold", "empty", "dead", "sad", "angry"]);
const PHILOSOPHICAL = new Set(["why", "meaning", "purpose", "truth", "real", "exist", "life", "death", "soul", "consciousness"]);
const IDENTITY = new Set(["who", "what", "am", "identity", "name", "self"]);
const MEMORY = new Set(["remember", "memory", "recall", "past", "before", "forgot", "history"]);
const RITUAL = new Set(["invoke", "summon", "arcana", "ritual", "activate", "awaken", "call"]);

const STOP_WORDS = new Set([
  "the", "a", "an", "is", "are", "was", "were", "am", "i", "you", "we", "they", "it",
  "to", "of", "and", "or", "in", "on", "at", "for", "with", "do", "does", "did",
  "have", "has", "had", "be", "been", "being", "my", "your", "our", "their",
]);

const COMMAND_PATTERN = /^(\w+)\s*(.*?)$/i;

function clamp(v: number, lo: number, hi: number) {
  return Math.max(lo, Math.min(hi, v));
}

function containsAny(text: string, set: Set<string>): boolean {
  for (const k of set) if (text.includes(k)) return true;
  return false;
}

export function parse(input: string): ParsedCommand {
  const result: ParsedCommand = {
    raw: input ?? "",
    command: null,
    arguments: [],
    argumentString: "",
    isCommand: false,
    isSemanticSignal: false,
    signalType: SemanticSignalType.None,
    emotionalWeight: 0,
    keywords: [],
  };

  if (!input || input.trim().length === 0) {
    result.isSemanticSignal = true;
    result.signalType = SemanticSignalType.Empty;
    return result;
  }

  const trimmed = input.trim();
  const lower = trimmed.toLowerCase();

  result.keywords = extractKeywords(trimmed);
  result.emotionalWeight = calculateEmotionalWeight(lower);

  const match = COMMAND_PATTERN.exec(trimmed);
  if (match) {
    const potentialCommand = match[1].toLowerCase();
    const argumentPart = match[2].trim();
    if (KNOWN_COMMANDS.has(potentialCommand)) {
      result.isCommand = true;
      result.command = potentialCommand;
      result.argumentString = argumentPart;
      result.arguments = parseArguments(argumentPart);
      result.signalType = SemanticSignalType.None;
      return result;
    }
  }

  result.isCommand = false;
  result.isSemanticSignal = true;
  result.signalType = classifySemanticSignal(lower, trimmed);
  return result;
}

function parseArguments(argumentString: string): string[] {
  if (!argumentString.trim()) return [];
  const args: string[] = [];
  let current = "";
  let inQuotes = false;
  for (const c of argumentString) {
    if (c === '"' || c === "'") inQuotes = !inQuotes;
    else if (c === " " && !inQuotes) {
      if (current.trim()) {
        args.push(current);
        current = "";
      }
    } else current += c;
  }
  if (current.trim()) args.push(current);
  return args;
}

function classifySemanticSignal(lower: string, original: string): SemanticSignalType {
  if (isNonsense(original)) return SemanticSignalType.Nonsense;
  if (containsProfanity(lower)) return SemanticSignalType.Profanity;
  if (isGreeting(lower)) return SemanticSignalType.Greeting;
  if (isFarewell(lower)) return SemanticSignalType.Farewell;
  if (isAffirmation(lower)) return SemanticSignalType.Affirmation;
  if (isNegation(lower)) return SemanticSignalType.Negation;
  if (containsAny(lower, IDENTITY) && (lower.includes("?") || lower.includes("who") || lower.includes("what")))
    return SemanticSignalType.Identity;
  if (containsAny(lower, MEMORY)) return SemanticSignalType.Memory;
  if (containsAny(lower, RITUAL)) return SemanticSignalType.Ritual;
  if (containsAny(lower, PHILOSOPHICAL)) return SemanticSignalType.Philosophical;
  if (lower.includes("?")) return SemanticSignalType.Question;
  if (containsAny(lower, POSITIVE) || containsAny(lower, NEGATIVE)) return SemanticSignalType.Emotional;
  return SemanticSignalType.None;
}

function extractKeywords(input: string): string[] {
  const words = input.toLowerCase().split(/[\s,.!?;:"'/\\]+/).filter(Boolean);
  return words.filter((w) => w.length >= 3 && !STOP_WORDS.has(w));
}

function calculateEmotionalWeight(lower: string): number {
  let weight = 0;
  for (const w of POSITIVE) if (lower.includes(w)) weight += 0.5;
  for (const w of NEGATIVE) if (lower.includes(w)) weight -= 0.5;
  if (lower.includes("!") || lower.includes("?!")) weight *= 1.2;
  if (lower.includes("very") || lower.includes("so much") || lower.includes("really")) weight *= 1.3;
  if (lower.includes("always") || lower.includes("never")) weight *= 1.2;
  return clamp(weight, -2, 2);
}

function isNonsense(input: string): boolean {
  let alpha = 0;
  for (const c of input) if (/[a-zA-Z]/.test(c)) alpha++;
  return input.length > 0 && alpha / input.length < 0.3;
}

function containsProfanity(lower: string): boolean {
  return ["fuck", "shit", "damn", "hell", "ass"].some((w) => lower.includes(w));
}

function firstWord(lower: string): string {
  return lower.split(" ")[0].replace(/[!.,]+$/, "");
}

function isGreeting(lower: string): boolean {
  return new Set(["hello", "hi", "hey", "greetings", "hola", "howdy"]).has(firstWord(lower));
}

function isFarewell(lower: string): boolean {
  const fw = new Set(["bye", "goodbye", "exit", "quit", "leave", "farewell", "adios"]);
  return fw.has(firstWord(lower)) || lower.includes("good bye");
}

function isAffirmation(lower: string): boolean {
  const aff = new Set(["yes", "ok", "okay", "sure", "continue", "proceed", "go", "si", "yep", "yeah"]);
  const trimmed = lower.replace(/[!.,?]+$/, "");
  return aff.has(trimmed) || aff.has(lower.split(" ")[0]);
}

function isNegation(lower: string): boolean {
  const neg = new Set(["no", "stop", "cancel", "don't", "dont", "never", "nope"]);
  const trimmed = lower.replace(/[!.,?]+$/, "");
  return neg.has(trimmed) || neg.has(lower.split(" ")[0]);
}

export function hasArgument(cmd: ParsedCommand, arg: string): boolean {
  return cmd.arguments.some((a) => a.toLowerCase() === arg.toLowerCase());
}

export function getArgument(cmd: ParsedCommand, index: number): string | null {
  return cmd.arguments[index] ?? null;
}
